uuv_dave/urdf/parallel_gripper.gazebo.xacro

```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0"?>
+
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
+    <xacro:macro name="parallel_gripper_gazebo_macro" params="namespace">
+        <gazebo reference="${namespace}/wrist">
+            <sensor name="arm_wrist" type="force_torque">
+                <force_torque>
+                <frame>child</frame>
+                <measure_direction>child_to_parent</measure_direction>
+                <always_on>true</always_on>
+                <update_rate>30</update_rate>
+                </force_torque>
+            </sensor>
+        </gazebo>
+
+        <gazebo reference="${namespace}/finger_left_joint">
+            <sensor name="arm_finger_left_joint" type="force_torque">
```

Comment:
@bsb808 I just want to make sure I am understanding this right. This sensor tag is completely independent of ROS. All it does is publish wrench messages to a gazebo topic. So what I need to do to get this to a ROS topic is to write a ROS-aware plugin that reads those force_toruqe messages from the gazebo topic and republish them to a ROS topic.  
Is my understanding correct?

Reply:
@Youssef-Khaky Yes, you have explained it well.  This is a typical design pattern for Gazebo/ROS.  Working with both can feel a bit like loving in parallel universes simultaneously.  

I imagine you saw this tutorial - http://gazebosim.org/tutorials?tut=force_torque_sensor&cat=sensors   The `gz topic` command is a good introspection tool to see the traffic, but not as well documented as `rostopic`

It looks like the `force_torque` gazebo sensor publishes a WrenchStamped gazebo message (protobuf) which you can see here - https://github.com/osrf/gazebo/tree/gazebo11/gazebo/msgs
I also typically take a quick look at the sensor code to better understand how it works - https://github.com/osrf/gazebo/tree/gazebo11/gazebo/sensors

There are at least two ROS plugins that do some of what you want.
 
1. The FT ROS plugin: https://github.com/ros-simulation/gazebo_ros_pkgs/blob/kinetic-devel/gazebo_plugins/src/gazebo_ros_ft_sensor.cpp .   This includes some noise.   Here is an example of using the sensor that looks reasonable: https://github.com/CentroEPiaggio/force-torque-sensor/blob/master/model/ft_sensor.gazebo.xacro  
2. F3D ground truth plugin: http://gazebosim.org/tutorials?tut=ros_gzplugins  and https://github.com/ros-simulation/gazebo_ros_pkgs/blob/kinetic-devel/gazebo_plugins/src/gazebo_ros_f3d.cpp  which doesn't include noise.

You might take a look at the code for 1 and 2 to verify that they are doing the same thing with the exception of adding noise.

One thing I find useful in planning this type of thing is to draw a quick "boxes and arrows" sketch of what is going on.  Here is an example from the sonar effort:  https://github.com/Field-Robotics-Lab/dave/blob/master/docs/sonar_dev_arch.png
I use yEd to create the sketches.  It is nice b/c it is text based and fits well into repos.  In addition to getting things straight in my mind it helps to coordinate work with other and to make it more maintainable if folks have an idea of the high-level structure.  

I think you could start with the existing gazebo/gazebo_ros functionality to add a sensor to each joint, but it would be nice to automate the system such that a user could specify force feedback for an entire manipulator model and then the system would create the appropriate gazebo/gazebo_ros plugins for all the joints.  Might be able to do this with xacro or embedded ruby?

You might take a look at the HAPTIX project and ROS Industrial to make sure this doesn't exist somewhere.  Carlos wasn't aware of anything, but I'm surprised that it it isn't around and want to avoid reinventing any wheels.  If indeed isn't done it will be a really nice contribution.

Reply:
These are good resources. Thank you.  
I will look into automating the entire manipulator arm's sensor attachment.

Reply:
I think I am more in favour of the first one because it is applied to joints were the second one is applied to links. From my understanding, these kind of sensors are placed at the joints.  
Additionally, applying them at the joints means they have torque, the second one doesn't have torque since it is on links.  
I looked at couple of these sensors. The common configeration I have been noticing is putting those FT sensors at the base of an end-effector and not on every joint. Like [this](https://youtu.be/K5Z_pmBNrP8?t=65) video from Robotiq and [this](https://youtu.be/b4nz_hAh7qs?t=126) video from ATI.  Should we consider having it only at on the wrist joint?

Reply:
I agree.  Using the FT sensors seems like the way to go.

I like the Robotiq demo video - I think doing something similar in gazebo would be really useful.  Having a spring to push against is a good idea!

We want to have feedback from all the joints.  The existing hardware gives feedback to teleoperators for each joint and we anticipate that will be a useful part of u/w manipulation where visibility is very limited.